src/libsyntax/parse/parser.rs

```diff
@@ -2908,12 +2907,29 @@ impl<'a> Parser<'a> {
                 }
             }
             None => {
-                if as_prec > min_prec && self.eat_keyword(keywords::As) {
+                if as_prec >= min_prec && self.eat_keyword(keywords::As) {
                     let rhs = self.parse_ty();
                     let _as = self.mk_expr(lhs.span.lo,
                                            rhs.span.hi,
                                            ExprCast(lhs, rhs));
                     self.parse_more_binops(_as, min_prec)
+                } else if range_prec >= min_prec
+                        && match lhs.node { ExprRange(_, _) => false, _ => true }
+                        && self.eat(&token::DotDot) {
+                    // '..' range notation, infix or postfix form
+                    // Note that we intentionally reject other range expressions on the lhs.
+                    // This makes '..1..2' invalid.
+                    // This is necessary for consistency between the prefix and postfix forms.
```

Comment:
So this means that things like `a..b..c..d` don't parse (at least not without parens) -- which is not necessarily bad, since that is nonsense, and what the user presumably wants is `(a..b)..(c..d)`, but I'd still sort of expect it to just produce a left- or right-leaning tree. I'm wondering if the goal of this rule was just to prevent such constructs from parsing, or if there was another concern? (If that is the goal, I'm not necessarily opposed, I guess it's consistent with `a < b < c` not parsing.)